File: kotlin/pig-latin/src/main/kotlin/PigLatin.kt
object PigLatin {

    private val vowels = arrayOf('a', 'e', 'i', 'o', 'u')

    fun translate(phrase: String): String {
        return phrase
            .split(" ")
            .joinToString(" ") { translateWord(it) }
    }

    private fun translateWord(word: String): String {
        val idx = firstConsonantClusterEnd(word)
        val quIndex = word.indexOf("qu")
        val yIndex = word.indexOf('y')

        return when {
            word[0] in vowels || word.startsWith("xr") || word.startsWith("yt") -> word + "ay"

            quIndex != -1 && quIndex < idx ->
                word.substring(quIndex + 2) +
                        word.substring(0, quIndex + 2) + "ay"

            yIndex > 0 && word.substring(0, yIndex)
                .all { it !in vowels } -> word.substring(yIndex) + word.substring(0, yIndex) + "ay"

            else -> word.substring(idx) + word.substring(0, idx) + "ay"
        }
    }

    private fun firstConsonantClusterEnd(word: String): Int {
        for (i in word.indices) {
            if (word[i] in vowels) return i
        }
        return word.length
    }
}

File: csharp/pig-latin/PigLatin.cs
public static class PigLatin
{
    private static readonly char[] vowels = ['a', 'e', 'i', 'o', 'u'];

    public static string Translate(string word) => string.Join(' ', word.Split(' ').Select(TranslateWord));

    private static string TranslateWord(string word)
    {
        var idx = FirstConsonantClusterEnd(word);
        var quIndex = word.IndexOf("qu");
        var yIndex = word.IndexOf('y');

        if (vowels.Contains(word[0]) || word.StartsWith("xr") || word.StartsWith("yt"))
        {
            return $"{word}ay";
        }
        else if (quIndex != -1 && quIndex < idx)
        {
            return $"{word[(quIndex + 2)..]}{word[..(quIndex + 2)]}ay";
        }
        else if (yIndex > 0 && word.Substring(0, yIndex).All(c => !vowels.Contains(c)))
        {
            return $"{word[yIndex..]}{word[..yIndex]}ay";
        }
        else
        {
            return $"{word[idx..]}{word[..idx]}ay";
        }
    }

    private static int FirstConsonantClusterEnd(string word)
    {
        for (var i = 0; i < word.Length; i++)
        {
            if (vowels.Contains(word[i])) return i;
        }

        return word.Length;
    }
}